### PR TITLE
MDEV-6284: Merge downstream Debian/Ubuntu packaging into upstream MariaDB

### PR DESCRIPTION
**Pull request by ottok**

### CODE REVIEW

**Comment by ottok** [comment]
This can be merged after https://github.com/MariaDB/server/pull/1463 and https://github.com/MariaDB/server/pull/1478 and https://github.com/MariaDB/server/pull/1453 have been merged first, otherwise there will be known CI verifiable upgrade/install/mtr errors.

File: debian/mariadb-server-10.5.mysql-server.logrotate [hunk 1]
```diff
@@ -11,7 +11,7 @@
 	sharedscripts
 	postrotate
           test -x /usr/bin/mysqladmin || exit 0
-          if [ -f `my_print_defaults --mysqld | grep -oP "pid-file=\K[^$]+"` ]; then
+          if [ -f `my_print_defaults --mysqld | grep -m 1 -oP "pid-file=\K.+$"` ]; then
             # If this fails, check debian.conf!
```
Comment (grooverdan) [inline debian/mariadb-server-10.5.mysql-server.logrotate:15]: Suggest using something like: https://github.com/MariaDB/server/pull/934#discussion_r340256147

File: debian/additions/mariadb-report [hunk 2]
```diff
@@ -147,8 +148,8 @@ my $use_thread_pool = 0;
 if(defined $op{'r'})
 {
    if($relative_live)
-   { 
-      print STDERR "mysqlreport is writing relative reports to '$tmpfile'.\n" unless $op{'detach'}; 
+   {
+      print STDERR "mysqlreport is writing relative reports to '$tmpfile'.\n" unless $op{'detach'};
```
Comment (grooverdan) [inline debian/additions/mariadb-report:152]: naming ^ now mariadbreport.

File: debian/not-installed [hunk 3]
```diff
@@ -1,7 +1,6 @@
 lib/systemd/system/mariadb.service # Installed by rules file
 lib/systemd/system/mariadb@.service # Installed by rules file
 usr/bin/mysql_config # We already have the MariaDB variant
-usr/bin/mysql_embedded # Huge 500 MB file. Not intended for distribution via any disto package.
 usr/bin/mytop # Mytop is distributed from a separate source package
 usr/bin/sst_dump # Use the one from rocksdb-tools package
 usr/lib/sysusers.d/sysusers.conf
```
Comment (grooverdan) [inline debian/not-installed:6]: why is sysusers.conf and tmpfile.conf not installed? Sure they need to be be mariadb.conf in the respective directories for naming. Any other reason?

Comment (cvicentiu) [reply debian/not-installed:6]: Actually in 742c36d048e0 you can configure exactly where sysusers.conf (named mariadb.conf) actually land. make install should put them in the right place. @ottok would this work for debian? That was my intent when merging that patch.

**Comment by ottok** [comment]
For this I also created some additional CI that actually tests the upgrade scenarios etc. Test run visible at https://salsa.debian.org/mariadb-team/mariadb-server/pipelines/124140/builds

File: debian/additions/mariadb-report (hunk 2 above)
Comment (ottok) [reply debian/additions/mariadb-report:152]: Renames extended

File: debian/not-installed (hunk 3 above)
Comment (ottok) [reply debian/not-installed:6]: This commit does not change that. I'll post separately on my init/systemd changes in another PR.

File: debian/mariadb-server-10.5.mysql-server.logrotate (hunk 1 above)
Comment (ottok) [reply debian/mariadb-server-10.5.mysql-server.logrotate:15]: Dropped this commit from this PR, let the other PR fix this instead once and for all.

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+# utf8 4-byte character set. See also client.cnf
+#
+character-set-server  = utf8mb4
+collation-server      = utf8mb4_general_ci
+
+#
+# * InnoDB
+#
+# InnoDB is enabled by default with a 10MB datafile in /var/lib/mysql/.
+# Read the manual for more InnoDB related options. There are many!
```
Comment (grooverdan) [inline debian/additions/mariadb.conf.d/50-server.cnf:111]: recommend including '#innodb_buffer_pool_size=8G' here. There are many innodb options however the default 128M is woefully inadequite in most cases. Sure people still use 1G database server (and complain about the performance), however 8G sends a message it should be big. Include the 80% of available ram recommendation or similar wording [from the kb](https://mariadb.com/kb/en/innodb-system-variables/#innodb_buffer_pool_size)

Comment (grooverdan) [reply debian/additions/mariadb.conf.d/50-server.cnf:111]: another quote from https://hackmongo.com/post/queries-effect-performance/ under 'Why not MySQL or hardware?'

> MySQL really only has one configuration that needs to be set properly: InnoDB buffer pool size

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+#
+# For generating SSL certificates you can use for example the GUI tool "tinyca".
+#
+#ssl-ca = /etc/mysql/cacert.pem
+#ssl-cert = /etc/mysql/server-cert.pem
+#ssl-key = /etc/mysql/server-key.pem
+#
+# Accept only connections using the latest and most secure TLS protocol version.
+# ..when MariaDB is compiled with OpenSSL:
+#ssl-cipher = TLSv1.2
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply debian/additions/mariadb.conf.d/50-server.cnf:94]: note MDEV-22221 - no definite conclusion at the moment.

**Comment by ottok** [comment]
Branch from https://github.com/MariaDB/server/pull/1498:
image

This branch:
image

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+#thread_stack           = 192K
+#thread_cache_size      = 8
+# This replaces the startup script and checks MyISAM tables if needed
+# the first time they are touched
+#myisam_recover_options = BACKUP
+#max_connections        = 100
+#table_cache            = 64
+#thread_concurrency     = 10
+
+#
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:44]: Removing the section would turn off query cache? From the page you linked:
> Starting from MariaDB 10.1.7, query_cache_type is automatically set to ON if the server is started with the query_cache_size set to a non-zero (and non-default) value.

If query cache is bad and should not be used, why is that not mentioned in the config docs you linked to?

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+#key_buffer_size        = 16M
+#max_allowed_packet     = 16M
+#thread_stack           = 192K
+#thread_cache_size      = 8
+# This replaces the startup script and checks MyISAM tables if needed
+# the first time they are touched
+#myisam_recover_options = BACKUP
+#max_connections        = 100
+#table_cache            = 64
+#thread_concurrency     = 10
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:42]: Thanks, removed

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+#general_log_file       = /var/log/mysql/mysql.log
+#general_log            = 1
+#
+# Error log - should be very few entries.
+#
+log_error = /var/log/mysql/error.log
+#
+# Enable the slow query log to see queries with especially long duration
+#slow_query_log_file    = /var/log/mysql/mariadb-slow.log
+#long_query_time        = 10
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:65]: True, but then again on a slow system you might have the whole system grinding to a halt and having everything logged over 1 second could create a lot of noice.

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+#log_slow_rate_limit    = 1000
+#log_slow_verbosity     = query_plan
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:67]: Updated

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+#log-queries-not-using-indexes
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:68]: done

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+#general_log            = 1
+#
+# Error log - should be very few entries.
+#
+log_error = /var/log/mysql/error.log
+#
+# Enable the slow query log to see queries with especially long duration
+#slow_query_log_file    = /var/log/mysql/mariadb-slow.log
+#long_query_time        = 10
+#log_slow_rate_limit    = 1000
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:66]: but rate limit of 1000 is quite a high number?

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+expire_logs_days        = 10
+#max_binlog_size        = 100M
+#binlog_do_db           = include_database_name
+#binlog_ignore_db       = exclude_database_name
+
+#
+# * Security Features
+#
+# Read the manual, too, if you want chroot!
+#chroot = /var/lib/mysql/
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:84]: removed

File: debian/additions/mariadb.conf.d/50-server.cnf (hunk 5 above)
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:94]: done

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+#
+#ssl-ca = /etc/mysql/cacert.pem
+#ssl-cert = /etc/mysql/server-cert.pem
+#ssl-key = /etc/mysql/server-key.pem
+#
+# Accept only connections using the latest and most secure TLS protocol version.
+# ..when MariaDB is compiled with OpenSSL:
+#ssl-cipher = TLSv1.2
+# ..when MariaDB is compiled with YaSSL (default in Debian):
+#ssl = on
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:96]: In mariadb.org OpenSSL is default and on Debian official WolfSSL is used.

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+#
+# The following can be used as easy to replay backup logs or for replication.
+# note: if you are setting up a replication slave, see README.Debian about
+#       other settings you may need to change.
+#server-id              = 1
+#log_bin                = /var/log/mysql/mysql-bin.log
+expire_logs_days        = 10
+#max_binlog_size        = 100M
+#binlog_do_db           = include_database_name
+#binlog_ignore_db       = exclude_database_name
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:78]: removed

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+# Instead of skip-networking the default is now to listen only on
+# localhost which is more compatible and is not less secure.
+bind-address            = 127.0.0.1
+
+#
+# * Fine Tuning
+#
+#key_buffer_size        = 16M
+#max_allowed_packet     = 16M
```
[the remark this answers is not in the document]
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:34]: done

File: debian/additions/mariadb.conf.d/50-server.cnf (hunk 4 above)
Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:111]: done

File: debian/additions/mariadb.conf.d/50-server.cnf (hunk 6 above)
[the remark this answers is not in the document]
Comment (grooverdan) [reply debian/additions/mariadb.conf.d/50-server.cnf]: Documentation should be updated. References MDEV-12200 and MDEV-18764

The MariaDB implementation if materially equivalent to the MySQL implementation, before they removed it in MySQL-8.0:
https://mysqlserverteam.com/mysql-8-0-retiring-support-for-the-query-cache/
https://dba.stackexchange.com/questions/217577/why-mysql-remove-query-cache-in-8-0-version

File: debian/additions/mariadb.conf.d/50-server.cnf (hunk 11 above)
[the remark this answers is not in the document]
Comment (grooverdan) [reply debian/additions/mariadb.conf.d/50-server.cnf:59]: so 1 in 1000 slow queries are logged. This means a less common slow query could be totally missed. With the default more queries will be logged, but logrotate is there, and its still a comment. Its usually to use tools like pt-query-digest to prioritize the queries of the slow query log.

File: debian/additions/mariadb.conf.d/50-server.cnf [hunk 16]
```diff
@@ -30,94 +29,80 @@ bind-address            = 127.0.0.1
...
 #ssl-cert = /etc/mysql/server-cert.pem
 #ssl-key = /etc/mysql/server-key.pem
-#
+
 # Accept only connections using the latest and most secure TLS protocol version.
-# ..when MariaDB is compiled with OpenSSL:
-#ssl-cipher = TLSv1.2
-# ..when MariaDB is compiled with YaSSL (default in Debian):
+# ..when MariaDB is compiled with OpenSSL (default in mariadb.org packaging):
+#ssl-cipher = TLSv1.3
```
Comment (grooverdan) [inline debian/additions/mariadb.conf.d/50-server.cnf:84]: https://mariadb.com/kb/en/ssltls-system-variables/#tls_version

I suspect openssl will allow the tls version in a cipher but wolfssl won't.

noting https://mariadb.com/kb/en/using-tlsv13/ might be approprate if you include ssl-cipher here still.

Comment (ottok) [reply debian/additions/mariadb.conf.d/50-server.cnf:84]: I'll just remove the ssl-cipher section and refer to https://mariadb.com/kb/en/securing-connections-for-client-and-server/

**Comment by ottok** [comment]
@grooverdan: I split out the config part into a separate issue: https://github.com/MariaDB/server/pull/1504

File: debian/rules [hunk 17]
```diff
@@ -128,14 +135,15 @@ endif
 	# Delete private files from libraries so they don't get shipped in the -dev pacakges
 	rm -r $(TMP)/usr/include/mariadb/server/private
 
+	# Don't ship sql-bench at all, just delete it completely even though it builds
+	rm -r $(TMP)/usr/sql-bench
```
Comment (cvicentiu) [inline debian/rules:139]: I am not aware of any reasons why this is removed. Was there a discussion on this that I missed?

We should probably do a proper cmake fix to allow one to *not* build this if we don't want to ship it.
Then again, if we don't ship it, why not have it in a separate repository?
Any other input you can give me regarding this so I understand it better?

**Comment by ottok** [comment]
I have no clue why sqlbench sources are embedded in mariadb repo and why it
is built. This rules line exists only to clean away the whole directory so
that the Debian QA tools can better handle comparing built vs. shipped
files.

We try to keep the difference small and list in debian/not-installed all
files that are built but not shipped on purpose. If we don't delete the
sqlbench directory, then the other option is to add it to
debian/not-installed.

**Comment by ottok** [comment]
Btw, this is already on buildbot:
http://buildbot.askmonty.org/buildbot/grid?branch=ok-10.5-merge-downstream&category=main&category=experimental

**Pull request merged**